### ISSUE TITLE
Add section on type conversion to rules-dsl.md

### ISSUE THREAD

**Issue by LightIsLife** [openhab/openhab-docs#472]
Suggested by @rkoshak in comments related to pull request #469, a section on type conversion needs to be added to the file configurations/rules-dsl.md.
Content of this section will be based on forum contribution https://community.openhab.org/t/type-conversions/32684 (also by @rkoshak). I am planning to develop a draft that can be discussed in this thread.

**Comment by ThomDietrich** [comment]
Hey! This thread is in fact the result of a discussion between the two of us and we decided to start it in the community. Later we want to transfer it to rules-dsl.md :)

**Comment by LightIsLife** [comment]
@ThomDietrich works for me, I am pretty agnostic as to where this discussion is best placed. @rkoshak suggested to raise an issue and discuss it here as a follow-up between the disucssion you and he had, you seem to prefer it in the forum, just let me know where you want to have this discussion and whether to should close this issue it what the desired etiquette is. Really not trying to be snippy, but not sure where to go from here.

**Comment by ThomDietrich** [comment]
There's really no "etiquette" for this, don't worry. I always go with what I learned over the years: Use whichever channel works. In order to discuss such a wide topic many non-github users might have an opinion on I think the community forum is the better place to discuss the initial version of the document. We'll see how it goes, so far there were not as many answers as I expected. I've left a comment inviting you to the game ;)

Anyhow, let's leave the issue open for later when we start to merge the article.

**Comment by rkoshak** [comment]
> @rkoshak suggested to raise an issue and discuss it here

Actually what I meant, though I admit it wasn't clear the way I worded it, was it might be worth creating the issue just to track the edits, but that the discussion should take place on the forum.

But I'll second Thom, it's all good. The main thing is to make some progress, no matter where it occurs.

**Comment by LightIsLife** [comment]
Thanks both, no problem....happy to work in whatever way is most convenient to you (as I have no preferred routine yet established :smile: ).  The forum it is to develop this issue

**Comment by LightIsLife** [comment]
Development here on doc section for type conversion: https://community.openhab.org/t/type-conversions/32684 is coming to a good end soon. @ThomDietrich I would like your guidance soon on how to proceed; PR #469 is still pending and this new section on type conversion should go immediately after the text proposed in #469 (just before the heading Scripts); will this create conflicting pull requests? Not super urgent as likely it will still be another week or so but wanted to start asking for guidance now.

**Comment by ThomDietrich** [comment]
Heyho, on vacation right now but I'm doing a pool GitHub session right now. Let's see how far I come.
In general I'd agree, wee can start the PR soon. I already left a comment regarding uniformity, maybe you could do something in the direction? I was expecting more users to chime in... anyhow.

**Comment by LightIsLife** [comment]
For reference, here the section that was deleted from #469. Just parking it here in the comments for now so it does not get lost as it was decided that it should be added to the type conversion text (this issue):

#### Details:

As all object-oriented computer languages, Java and the Rules DSL have implemented the concept of inheritance. 
However, inheritance only applies to Objects and does **not** apply to primitives. 
Inheritance allows to take an existing Object type, called a Class, and adding to it to make it into something different. 
This “something different” becomes a Child of the original Class, the parent. The Child still can do everything the parent could do. 
The top level base Class for all Objects in Java and the Rules DSL is called simply `Object`. 

In addition to other useful things, the class `Object` implements a method called `toString`. 
And since `Object` is the parent of all Objects, ALL Classes also implement a `toString` method. 
_However primitives do not inherit from Object. 
They don't inherit from anything and they don't have any methods at all which includes the lack of a toString Method._

The `sendCommand` is a generic action and needs to be able to work with all Item types. 
Actions only support two String arguments as all Objects will support the conversion `toString`. 
`sendCommand (MyItem, new_state)` will automatically use the `MyItem.toString` method to convert MyItem into a String. 
It will also attempt to do so with the second argument if `new_state` is not already a String. 
However, if the second argument is a primitive, and not an Object, it does not carry a method `toString`. 
Thus, Rules DSL will not be able to cast `new_state` as a String. 
As a consequence, the use of `sendCommand(MyItem, primitive)`, using a primitive as the second argument, will almost always fail. 

The different syntax for the generic and the objective-specific differs and is given in the table below:

| Generic (Action)                 | Specific (Method)               |
|----------------------------------|---------------------------------|
| `postUpdate(MyItem, new_state)`  | `MyItem.postUpdate(new_state)`  |
| `sendCommand(MyItem, new_state)` | `MyItem.sendCommand(new_state)` |

`MyTimes.sendCommand()`, however, will use the `sendCommand` method that is suitable to make the necessary type conversions.
For example, the `NumberItem` class would have a `sendCommand(int)`, `sendCommand(long)`, `sendCommand(float)`, `sendCommand(double)`, `sendCommand(Number)`, `sendCommand(DecimalType)`, and `sendCommand(String)` method. 
Each of these separate methods is individually written to handle all of these different types of Objects. 
MyItem will automatically apply the method that corresponds to the argument type.

In a nutshell, using the syntax `MyItem.sendCommand(new_state)` or `MyItem.sendUpdate(new_state)` will help avoid many problems.
